### PR TITLE
[SPARK-35043][SQL] Add condition lambda and rule id to the resolve function family

### PR DESCRIPTION
**Pull request by sigmod**
### What changes were proposed in this pull request?

This PR contains:
- AnalysisHelper changes to allow the resolve function family to stop earlier without traversing the entire tree;
- Example changes in a few rules to support such pruning, e.g., ResolveRandomSeed, ResolveWindowFrame, ResolveWindowOrder, and ResolveNaturalAndUsingJoin.

### Why are the changes needed?

It's a framework-level change for reducing the query compilation time.
In particular, if we update existing analysis rules' call sites as per the examples in this PR, the analysis time can be reduced as described in the [doc](https://docs.google.com/document/d/1SEUhkbo8X-0cYAJFYFDQhxUnKJBz4lLn3u4xR2qfWqk).

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

It is tested by existing tests.

### CODE REVIEW

**Comment by sigmod** [comment]
@hvanhovell @gengliangwang @maryannxue @dbaliafroozeh  this PR is ready for review. Let me know if you have any questions. Thanks!

**Comment by gengliangwang** [comment]
@sigmod Thanks for the work!
I have left a comment in https://issues.apache.org/jira/browse/SPARK-34916?focusedCommentId=17319839&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17319839
Could you create an umbrella Jira for the framework and related works? Then, we can create a new JIRA for this PR.

**Comment by sigmod** [comment]
> @sigmod Thanks for the work!
> I have left a comment in https://issues.apache.org/jira/browse/SPARK-34916?focusedCommentId=17319839&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17319839
> Could you create an umbrella Jira for the framework and related works? Then, we can create a new JIRA for this PR.

Done. 
Thanks, Gengliang! I updated the JIRA issue in this PR to subtask SPARK-35043. Let me know if I missed anything.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -39,6 +39,9 @@ import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.rules._
 import org.apache.spark.sql.catalyst.streaming.StreamingRelationV2
 import org.apache.spark.sql.catalyst.trees.TreeNodeRef
+import org.apache.spark.sql.catalyst.trees.TreePattern.{
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:42]: Does this pass scalastyle?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala [hunk 2]
```diff
@@ -104,17 +155,30 @@ trait AnalysisHelper extends QueryPlan[LogicalPlan] { self: LogicalPlan =>
...
-    if (!analyzed) {
+    resolveOperatorsDownWithPruning(AlwaysProcess.fn, UnknownRuleId)(rule)
+  }
+
+  /** Similar to [[resolveOperatorsUpWithPruning]], but does it top-down. */
+  def resolveOperatorsDownWithPruning(cond: TreePatternBits => Boolean,
+    ruleId: RuleId = UnknownRuleId)(rule: PartialFunction[LogicalPlan, LogicalPlan])
+  : LogicalPlan = {
+    if (!analyzed && cond.apply(self) && !isRuleIneffective(ruleId)) {
       AnalysisHelper.allowInvokingTransformsInAnalyzer {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala:166]: A bit of OCD, but do we really have to do this check for each recursive call? It seems a bit wasteful.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:42]: It seemed ok -- I didn't see scalastyle complained anything. 
But let me know if there's a convention in the Spark codebase.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala (hunk 2 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala:166]: Yeah, makes sense.
I'll see if I can remove allowInvokingTransformsInAnalyzer in a subsequent PR.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala:166]: Sounds good!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:42]: Normally we put it all on a single line. If it passes then I am happy.

**Review by gengliangwang — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:42]: It passes..

**Comment by gengliangwang** [comment]
GA passes. Merging to master.

**Pull request closed without merging**